src/main/java/org/folio/services/migration/item/ItemShelvingOrderMigrationService.java

```diff
@@ -0,0 +1,97 @@
+package org.folio.services.migration.item;
+
+import static org.apache.logging.log4j.LogManager.getLogger;
+import static org.folio.dbschema.ObjectMapperTool.readValue;
+
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.sqlclient.Row;
+import io.vertx.sqlclient.RowStream;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.Logger;
+import org.folio.rest.jaxrs.model.Item;
+import org.folio.rest.persist.PgUtil;
+import org.folio.rest.persist.PostgresClientFuturized;
+import org.folio.rest.support.EffectiveCallNumberComponentsUtil;
+import org.folio.services.item.ItemService;
+import org.folio.services.migration.BaseMigrationService;
+import org.folio.services.migration.BatchedReadStream;
+
+public class ItemShelvingOrderMigrationService extends BaseMigrationService {
+  private static final Logger log = getLogger(ItemShelvingOrderMigrationService.class);
+  private static final String SELECT_SQL = "SELECT jsonb FROM %s WHERE "
+    + "jsonb->>'effectiveShelvingOrder' IS NULL";
+
+  private final PostgresClientFuturized postgresClient;
+  private final ItemService itemService;
+
+  public ItemShelvingOrderMigrationService(Context context, Map<String, String> okapiHeaders) {
+    this(new PostgresClientFuturized(PgUtil.postgresClient(context, okapiHeaders)),
+      new ItemService(context, okapiHeaders));
+  }
+
+  public ItemShelvingOrderMigrationService(
+    PostgresClientFuturized postgresClient, ItemService itemService) {
+
+    super("20.2.0");
+    this.postgresClient = postgresClient;
+    this.itemService = itemService;
+  }
+
+  @Override
+  public Future<Void> runMigration() {
+    log.info("Starting item shelving order migration... ");
+
+    return postgresClient.startTx()
+      .compose(con -> postgresClient.selectStream(con, selectSql())
+        .compose(this::handleUpdate)
+        .onSuccess(records -> log.info(
+          "Shelving order migration has been completed [recordsProcessed={}]", records))
+        .onFailure(error -> log.error("Unable to complete shelving order migration", error))
+        .onComplete(result -> postgresClient.endTx(con)))
+      .mapEmpty();
+  }
+
+  private Future<Integer> handleUpdate(RowStream<Row> stream) {
+    var batchStream = new BatchedReadStream<>(stream);
+    var promise = Promise.<Integer>promise();
+    var recordsUpdated = new AtomicInteger(0);
+
+    batchStream
+      .endHandler(notUsed -> promise.tryComplete(recordsUpdated.get()))
+      .exceptionHandler(promise::tryFail)
+      .handler(rows -> {
+        var items = rows.stream()
+          .map(this::rowToItem)
+          .map(EffectiveCallNumberComponentsUtil::calculateAndSetEffectiveShelvingOrder)
+          .collect(Collectors.toList());
+
+        // Pause stream, so that updates is executed in sequence
+        batchStream.pause();
```

Comment:
Why is the stream paused here rather than before the items are semi-processed?

If the goal is for only one handler to be invoked at a time, shouldn't the batch be paused before the handler is invoked or at least the start of the handler (which would leave a race condition if the stream continues, would it not)?

Reply:
As far as I understand the stream never processed concurrently.

The `pause` is needed in order to wait for update of items to be finished, if we don't pause the stream than updates will be submitted and stream finished without confirming that updates succeeded/failed.

Reply:
> As far as I understand the stream never processed concurrently.

If the stream never invokes multiple handlers concurrently, how could the updates be executed out of sequence (referred to in the comment)?

> The pause is needed in order to wait for update of items to be finished, if we don't pause the stream than updates will be submitted and stream finished without confirming that updates succeeded/failed.

I don't think I follow. The underlying stream needs to be paused because if that stream finishes before the batched stream finishes processing then updates will be lost?

Reply:
> I don't think I follow. The underlying stream needs to be paused because if that stream finishes before the batched stream finishes processing then updates will be lost?

No, it is not related to the batched stream. Let's say we have following code:

```java
public User updateUser(User user) {
  updateUserLoginInfo(user.getLoginInfo());
  persistUser(user);

  return user;
}

private Future<Void> updateUserLoginInfo(UserLoginInfo login);
private Future<Void> persistUser(User user);
```

When we do something `var updatedUser = updateUser(user)` then at the time when we returned from the method the user most likely is not persist yet nor login info is updated because these two methods returns future, that means that result will be available after some while once the future is completed, right?

But if we still use this approach we can not guarantee that update is successful. 

That's why I pause/resume stream: I pause the stream before starting update, than wait for computable future is completed and depending on the result I resume or stop the stream.

Reply:
> When we do something var updatedUser = updateUser(user) then at the time when we returned from the method the user most likely is not persist yet nor login info is updated because these two methods returns future, that means that result will be available after some while once the future is completed, right?

Sure

> But if we still use this approach we can not guarantee that update is successful.

If we use what approach?

> That's why I pause/resume stream: I pause the stream before starting update, then wait for computable future is completed and depending on the result I resume or stop the stream.

I'm still not sure I understand why this is important.

I understand the need to use a continuation of the future to decide on whether the process is a success or not.

Why does it matter whether the stream has been paused? If no additional handlers are invoked until this one finishes, all that will happen is the stream will grow. Is that the case? Is that the undesirable behaviour?

Reply:
> If no additional handlers are invoked until this one finishes, all that will happen is the stream will grow.

The think is that update operations are handled on vertx thread pool, so once the `updateItem` is called handler will immediately continue their processing.

At some point the stream will be finished but some item updates will be still pending. I'm not sure if it is allowed behavior or not.

For reindex we don't wait and pause stream, and no memory issues observed.

Reply:
> That is the single handler that processes a batch of records?
Yes.

> Then why pause the stream here and not in that process?

Un which process, item update?

Reply:
> Un which process, item update?

I thought you said that the Re-index process does not pause the underlying stream during it's processing, yet this one (to update items) does. Did I misunderstand?

Reply:
Reindex is fully asynchronous and it is processed in background.

If it is acceptable that the stream is processed and tenant init is marked as completed, even though there might be some pending updates, than I'm happy to remove pause/resume logic.

Reply:
> Reindex is fully asynchronous and it is processed in background.

From your explanation, the purpose of the pause is to try to limit the memory usage during the process (rather than the ordering, which the comment referred to). I don't understand how whether the process runs synchronously or asynchronously relates to that.

I also thought that the Tenant API 2.0 runs asynchronously by default (it can be forced synchronous by a flag), with Okapi polling a status endpoint. Is that not the case?

> If it is acceptable that the stream is processed and tenant init is marked as completed, even though there might be some pending updates, than I'm happy to remove pause/resume logic.

I wasn't suggesting that the pause be removed. I was trying to understand why it was needed, as I couldn't understand the ordering comment when you also stated that only one handler instance is executed concurrently.

Anyway, given the meeting that @SergiyVSergiyenko has raised for this morning. I should probably stop asking questions. Please feel free to ignore.